libbpf-tools/schedsnoop.c

```diff
@@ -0,0 +1,309 @@
+// SPDX-Licence-Identifier: GPL-2.0
+// Copyright (c) 2020 Alibaba Cloud
+//
+// Based on schedsnoop by Michael Wang
+// Created by Wenhao Qu
+//
+// Maintainers:
+// Michael Wang <yun.wang@linux.alibaba.com>
+// Wenhao Qu <quxi.qwh@alibaba-inc.com>
+#include <argp.h>
+#include <stdio.h>
+#include <signal.h>
+#include <sched.h>
+#include <string.h>
+#include <unistd.h>
+#include <fcntl.h>
+#include <locale.h>
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <sys/time.h>
+#include <sys/resource.h>
+#include <sys/wait.h>
+#include <dirent.h>
+#include <bpf/bpf.h>
+#include "errno_helpers.h"
+#include "trace_helpers.h"
+#include "syscall_helpers.h"
+#include "schedsnoop.h"
+#include "schedsnoop.skel.h"
+
+#define NS_IN_SEC		1000000000LLU
+#define NS_IN_MS		1000000LLU
+#define NS_IN_US		1000LLU
+#define PERF_BUFFER_PAGES	64
+
+const char *argp_program_version = "schedsnoop 0.1";
+const char *argp_program_bug_address = "<bpf@vger.kernel.org>";
+static const char argp_program_doc[] =
+"Trace the related schedule events of a specified task.\n"
+"\n"
+"USAGE: schedsnoop -p PID [-s]\n"
+"\n"
+"EXAMPLES:\n"
+"    schedsnoop -p 49870       	# trace pid 49870\n"

```

Comment:
Tracing all threads of a given process is a great idea and we are working on it. Currently, if our target process has multiple threads, we will open multiple consoles and use this tool to trace each threads respectively.

A small problem is, how to show the traceinfos? Simply output them to the console is not so good. Even now when we trace one single thread, it output plenty of information to the console. When we  trace multiple threads at the same time, it will be a big chaos.

Our idea is that we only output very little information to the console, such as how many threads are we tracing, which of them are working and which of them are preempted. Detail information will be written to certain files related to each threads. Do you think it is ok? Or could you please give us some suggestions.

Reply:
dumping to file to visualize with some other tool is definitely an approach. Not sure how suitable it is for BCC tools, though if you have an option of outputting, e.g., JSON events, then you can build a pipeline that would parse it and dump in whatever closed/proprietary format is suitable for your visualization tool.

I don't think you need to implement any of that in the first version. More important to get right from the get go is to establishing good terminology: use tid for "thread ID", use "thread" or "task" terminology throughout, because that's what you are tracing.